src/Symfony/Component/Security/Http/LoginLink/LoginLinkHandler.php

```diff
@@ -0,0 +1,115 @@
+
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\PropertyAccess\PropertyAccessorInterface;
+use Symfony\Component\Routing\Generator\UrlGeneratorInterface;
+use Symfony\Component\Security\Core\Exception\UsernameNotFoundException;
+use Symfony\Component\Security\Core\User\UserInterface;
+use Symfony\Component\Security\Core\User\UserProviderInterface;
+use Symfony\Component\Security\Http\LoginLink\Exception\ExpiredLoginLinkException;
+use Symfony\Component\Security\Http\LoginLink\Exception\InvalidLoginLinkException;
+
+/**
+ * @author Ryan Weaver <ryan@symfonycasts.com>
+ * @experimental in 5.2
+ */
+final class LoginLinkHandler implements LoginLinkHandlerInterface
+{
+    private $urlGenerator;
+    private $userProvider;
+    private $propertyAccessor;
+    private $signatureProperties;
+    private $secret;
+    private $options;
+    private $expiredStorage;
+
+    public function __construct(UrlGeneratorInterface $urlGenerator, UserProviderInterface $userProvider, PropertyAccessorInterface $propertyAccessor, array $signatureProperties, string $secret, array $options, ?ExpiredLoginLinkStorage $expiredStorage)
+    {
+        $this->urlGenerator = $urlGenerator;
+        $this->userProvider = $userProvider;
+        $this->propertyAccessor = $propertyAccessor;
+        $this->signatureProperties = $signatureProperties;
+        $this->secret = $secret;
+        $this->options = array_merge([
+            'route_name' => null,
+            'lifetime' => 600,
+            'max_uses' => null,
+        ], $options);
+        $this->expiredStorage = $expiredStorage;
+    }
+
+    public function createLoginLink(UserInterface $user): LoginLinkDetails
+    {
+        $expiresAt = new \DateTimeImmutable(sprintf('+%d seconds', $this->options['lifetime']));
+
+        $expires = $expiresAt->format('U');
+        $parameters = [
+            'user' => $user->getUsername(),
+            'expires' => $expires,
+            'hash' => $this->computeSignatureHash($user, $expires),
+        ];
+
+        $url = $this->urlGenerator->generate(
+            $this->options['route_name'],
+            $parameters,
+            UrlGeneratorInterface::ABSOLUTE_URL
+        );
+
+        return new LoginLinkDetails($url, $expiresAt);
+    }
+
+    public function consumeLoginLink(Request $request): UserInterface
+    {
+        $username = $request->get('user');
+
+        try {
+            $user = $this->userProvider->loadUserByUsername($username);
+        } catch (UsernameNotFoundException $exception) {
+            throw new InvalidLoginLinkException('User not found.', 0, $exception);
+        }
+
+        $hash = $request->get('hash');
+        $expires = $request->get('expires');
+        if (false === hash_equals($hash, $this->computeSignatureHash($user, $expires))) {
+            throw new InvalidLoginLinkException('User has changed since link was sent.');
+        }
+
+        if ($expires < time()) {
```

Comment:
We could move this condition before the one which compares hashes (because it requires less computing and because "expired links" could be a common case ... maybe it's not, then ignore my comment)

Reply:
The order of these can be tricky for things like timing attacks. I'm following `TokenBasedRememberMeServices` exactly:

* load user object
* check hash
* check expires

So an entirely valid question, but I think it's correct how it is